tests/test_spyre_cb.py

```diff
@@ -0,0 +1,282 @@
+                                      temperature=0.0,
+                                      stop="1",
+                                      ignore_eos=True)
+
+    sampling_params3 = SamplingParams(max_tokens=max_tokens3,
+                                      temperature=0.0,
+                                      stop="1",
+                                      ignore_eos=True)
+
+    vllm_sampling_params = [
+        sampling_params1,
+        sampling_params2,
+        sampling_params3,
+    ]
+
+    # These prompts are ordered so that they don't finish in the
+    # order given
+    prompts = [
+        "7 6 5 4",
+        "10 9 8 7",
+        "8 7 6 5",
+    ]
+
+    # Ensure that both:
+    # - The model doesn't crash
+    # - The output sequences are correct
+    vllm_results = generate_cb_spyre_vllm_output(
+        model=model,
+        prompts=prompts,
+        max_model_len=2048,
+        block_size=2048,
+        sampling_params=vllm_sampling_params,
+        tensor_parallel_size=1,
+        backend=backend,
+    )
+
+    assert vllm_results[0]["text"] == " 3 2 "
+    assert vllm_results[1]["text"] == " 6 5 4 3 2 "
+    assert vllm_results[2]["text"] == " 4 3 2 "
+
+
+@pytest.mark.parametrize("model", get_spyre_model_list())
+@pytest.mark.parametrize("backend", ["eager", "inductor"])
+def test_cb_with_engine_steps(model: str, backend: str,
+                              monkeypatch: pytest.MonkeyPatch):
+    """Test that the spyre worker correctly handles 
+    continuous batches of requests and one sequence 
+    an exit the batch while a new ones gets prefilled 
+    and appended to the batch"""
+    with monkeypatch.context() as m:
+        max_tokens1 = 20
+        max_tokens2 = 17
+
+        sampling_params1 = SamplingParams(max_tokens=max_tokens1,
+                                          temperature=0.0,
+                                          stop="1",
+                                          ignore_eos=True)
+
+        sampling_params2 = SamplingParams(max_tokens=max_tokens2,
+                                          temperature=0.0,
+                                          stop="1",
+                                          ignore_eos=True)
+
+        prompt1 = "7 6 5 4"
+        prompt2 = "10 9 8 7"
+
+        # set env vars
+        m.setenv("VLLM_SPYRE_WARMUP_PROMPT_LENS", "64")
+        m.setenv("VLLM_SPYRE_WARMUP_NEW_TOKENS",
+                 str(max([max_tokens1, max_tokens2])))
+
+        m.setenv("VLLM_SPYRE_USE_CB", "1")
+        m.setenv("VLLM_USE_V1", "1")
+
+        m.setenv("VLLM_SPYRE_MAX_CONTEXT_LENGTH", "2048")
+        m.setenv("VLLM_SPYRE_MAX_BATCH_SIZE", str(2))
+        m.setenv("VLLM_SPYRE_DYNAMO_BACKEND", backend)
+
+        # start the engine
+        engine_args = EngineArgs(
+            model=model,
+            enforce_eager=True,  # reduce test time
+        )
+
+        engine = V1LLMEngine.from_engine_args(engine_args)
```

Comment:
Just thinking aloud: if you set `envs.VLLM_ENABLE_V1_MULTIPROCESSING = False`, `from_engine_args()` should create an engine that has an `InprocClient` for the engine core and then the engine step will be in lockstep with the engine core step. So you could combine both tests if you want. Although perhaps it would make the combined test a little harder to read.

Reply:
Hmm, yeah it's hard for me already to wrap my head around that :) Let me wrap up the basic tests first then I can try to take a look at this if that works